core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/DefaultAwsV4HttpSigner.java

```diff
@@ -72,157 +75,225 @@ public SyncSignedRequest sign(SyncSignRequest<? extends AwsCredentialsIdentity>
 
-        String host = requestBuilder.host();
-        if (!SdkHttpUtils.isUsingStandardPort(requestBuilder.protocol(), requestBuilder.port())) {
-            host += ":" + requestBuilder.port();
+        if (StringUtils.isNotBlank(checksumHeaderName) && checksumAlgorithm == null) {
+            throw new IllegalArgumentException(
+                CHECKSUM_ALGORITHM + " cannot be null when " + CHECKSUM_HEADER_NAME + " is given!");
         }
-        requestBuilder.putHeader(SignerConstant.HOST, host);
-    }
 
-    private void addDateHeader(SdkHttpRequest.Builder requestBuilder, String dateTime) {
-        requestBuilder.putHeader(SignerConstant.X_AMZ_DATE, dateTime);
+        return createSdkChecksumFromRequest(request.request(), checksumHeaderName, checksumAlgorithm);
     }
 
-    private void putChecksumHeader(SdkChecksum sdkChecksum,
-                                   SdkHttpRequest.Builder requestBuilder, String contentHashString, String checksumHeaderName) {
+    /**
+     * Generate a {@link ContentChecksum} from the {@link SyncSignRequest}.
+     */
+    protected ContentChecksum createChecksum(SyncSignRequest<?> request) {
+        SdkChecksum sdkChecksum = createSdkChecksum(request);
+        String contentHash = createContentHash(request.payload().orElse(null), sdkChecksum);
 
-        if (sdkChecksum != null && !isUnsignedPayload(contentHashString)) {
-            requestBuilder.putHeader(checksumHeaderName, BinaryUtils.toBase64(sdkChecksum.getChecksumBytes()));
-        }
+        return new ContentChecksum(contentHash, sdkChecksum);
     }
 
-    private Boolean isUnsignedPayload(String contentHashString) {
-        return "UNSIGNED_PAYLOAD".equals(contentHashString) || "STREAMING-UNSIGNED-PAYLOAD-TRAILER".equals(contentHashString);
+    /**
+     * Generate a {@link CompletableFuture<ContentChecksum>} from the {@link AsyncSignRequest}
+     */
+    protected CompletableFuture<ContentChecksum> createChecksum(AsyncSignRequest<?> request) {
+        SdkChecksum sdkChecksum = createSdkChecksum(request);
+
+        return createContentHash(request.payload().orElse(null), sdkChecksum).thenApply(
+            hash -> new ContentChecksum(hash, sdkChecksum));
     }
 
-    @Override
-    public AsyncSignedRequest signAsync(AsyncSignRequest<? extends AwsCredentialsIdentity> request) {
-        if (CredentialUtils.isAnonymous(request.identity())) {
-            return AsyncSignedRequest.builder()
-                .request(request.request())
-                .payload(request.payload().orElse(null))
-                .build();
-        }
+    /**
+     * Add any prerequisite items to the request using the {@link SdkHttpRequest.Builder}, the ${@link SignRequest},
+     * and the {@link ContentChecksum}
+     * <p>
+     * Such an item could be a header or query parameter that should be included in the signature of the request.
+     */
+    protected void addPrerequisites(SdkHttpRequest.Builder requestBuilder,
```

Comment:
nit: code layout. these methods in the middle of methods related to checksum made it slightly worse to follow the code while reading. co-locating related methods helps.